utils/service_discovery/abstract_config_store.py

```diff
@@ -0,0 +1,173 @@
+        self.client = None
+        self.agentConfig = agentConfig
+        self.settings = self._extract_settings(agentConfig)
+        self.client = self.get_client()
+        self.sd_template_dir = agentConfig.get('sd_template_dir')
+        self.AUTO_CONF_IMAGES = get_auto_conf_images(agentConfig)
+
+    @classmethod
+    def _drop(cls):
+        """Drop the config store instance"""
+        del cls._instances[cls]
+
+    def _extract_settings(self, config):
+        raise NotImplementedError()
+
+    def get_client(self, reset=False):
+        raise NotImplementedError()
+
+    def client_read(self, path, **kwargs):
+        raise NotImplementedError()
+
+    def dump_directory(self, path, **kwargs):
+        raise NotImplementedError()
+
+    def _get_auto_config(self, image_name):
+        if image_name in self.AUTO_CONF_IMAGES:
+            check_name = self.AUTO_CONF_IMAGES[image_name]
+
+            # get the check class to verify it matches
+            check = get_check_class(self.agentConfig, check_name)
+            if check is None:
+                log.info("Could not find an auto configuration template for %s."
+                         " Leaving it unconfigured." % image_name)
+                return None
+
+            auto_conf = get_auto_conf(self.agentConfig, check_name)
+            init_config, instances = auto_conf.get('init_config'), auto_conf.get('instances')
+
+            # stringify the dict to be consistent with what comes from the config stores
+            init_config_tpl = json.dumps(init_config) if init_config else '{}'
+            instance_tpl = json.dumps(instances[0]) if instances and len(instances) > 0 else '{}'
+
+            return (check_name, init_config_tpl, instance_tpl)
+        return None
+
+    def get_check_tpl(self, image, **kwargs):
+        """Retrieve template config strings from the ConfigStore."""
+        # this flag is used when no valid configuration store was provided
+        trace_config = kwargs.get('trace_config', False)
+        source = ''
+
+        if kwargs.get('auto_conf') is True:
+            auto_config = self._get_auto_config(image)
+            if auto_config is not None:
+                check_name, init_config_tpl, instance_tpl = auto_config
+                source = CONFIG_FROM_AUTOCONF
+            else:
+                log.debug('No auto config was found for image %s, leaving it alone.' % image)
+                return None
+        else:
+            try:
+                # Try to read from the user-supplied config
+                check_name = self.client_read(path.join(self.sd_template_dir, image, 'check_name').lstrip('/'))
+                init_config_tpl = self.client_read(path.join(self.sd_template_dir, image, 'init_config').lstrip('/'))
+                instance_tpl = self.client_read(path.join(self.sd_template_dir, image, 'instance').lstrip('/'))
+                source = CONFIG_FROM_TEMPLATE
+            except (KeyNotFound, TimeoutError):
```

Comment:
You should differentiate these 2 cases and add more specific logging to give better information about why it failed.